### PR TITLE
fix: reorder handling of `Response` replies

### PR DESCRIPTION
**Pull request by barbieri**
We must add `Response.headers` before proceeding, otherwise early returns (ie: `204`) will miss them.

We can also remove special handling of `ReadableStream`: since the `Response` usage is fully finished, we can continue with `Response.body` as the payload and the possible types `null` or `ReadableStream` will be handled by existing cases.

Last but not least, we should add `304` handling to the same early return handling `204`, including the removal of `content-*` headers.

> **NOTE:** this should be back ported to other branches, at least v4 suffers from the same issue.

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] ~documentation is changed or added~ (not needed)
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Review by gurgunday — APPROVED** [review]
lgtm

File: lib/reply.js [hunk 1]
```diff
@@ -627,7 +643,7 @@ function onSendEnd (reply, payload) {
     return
   }
 
-  if ((statusCode >= 100 && statusCode < 200) || statusCode === 204) {
+  if ((statusCode >= 100 && statusCode < 200) || statusCode === 204 || statusCode === 304) {
```
Comment (gurgunday) [inline lib/reply.js:646]: Does this count as a bugfix/feat or a breaking change?

Comment (barbieri) [reply lib/reply.js:646]: Bug fix, it was broken before.

Actually I started with this fix I was returning 304 and it was not working, but 204 was.

Then noticed the other problems: any `Response.body === null` would never end the reply and headers were not properly set in the body-less responses (including 204)

**Review by metcoder95 — APPROVED** [review]

**Review by climba03003 — APPROVED** [review]

**Comment by climba03003** [comment]
> who is responsible for that?

The label is used to activate workflows when this PR is merged.
So, no human works need to be done unless the workflows failed.

**Comment by mcollina** [comment]
This seems to be breaking fastify-etag tests.

File: lib/reply.js (hunk 1 above)
Comment (climba03003) [inline lib/reply.js:646]: Let revert the changes on `304` and land the re-ordering.
```suggestion
  if ((statusCode >= 100 && statusCode < 200) || statusCode === 204) {
```

**Review by climba03003 — CHANGES_REQUESTED** [review]

File: lib/reply.js (hunk 1 above)
Comment (barbieri) [reply lib/reply.js]: anyway, the PR is updated without this since the https://www.rfc-editor.org/rfc/rfc9110.html#section-8.6-7 states:

> A server MAY send a Content-Length header field in a [304 (Not Modified)](https://www.rfc-editor.org/rfc/rfc9110.html#status.304) response to a conditional GET request ([Section 15.4.5](https://www.rfc-editor.org/rfc/rfc9110.html#status.304)); a server MUST NOT send Content-Length in such a response unless its field value equals the decimal number of octets that would have been sent in the content of a [200 (OK)](https://www.rfc-editor.org/rfc/rfc9110.html#status.200) response to the same request

**Review by climba03003 — APPROVED** [review]
The error from CITGM is about the typescript or others.
Not related this changes.

**Review by gurgunday — APPROVED** [review]
lgtm

**Review by jsumners — APPROVED** [review]

**Comment by barbieri** [comment]
Thanks! Do you know when the next release is due?
